### PR TITLE
Refactor ABI Comparison

### PR DESCRIPTION
**Pull request by paytonrules**
When comparing two ABIs, save off the reason and give a good display message. Previously the panic would just show all the ABIs, and say "don't match".

Now you get, for example:

panic: ABIs don't match for contracts: MEDIAN_BAL and MEDIAN_BAT. Reason: constructors don't match constructorOne: constructor() returns(), constructorTwo:

### CODE REVIEW

File: pkg/shared/constants/external_test.go [hunk 1]
```diff
@@ -30,6 +31,59 @@ var _ = Describe("External constants", func() {
...
+
+		It("has an outer method error if the ABIs have the same constructor but differing outer methods", func() {
+			medianBalABI, parseErr := eth.ParseAbi(constants.GetContractABI("MEDIAN_BAL"))
+			Expect(parseErr).NotTo(HaveOccurred())
+			medianBalWithoutDenyABI, parseErr := eth.ParseAbi(constants.GetContractABI("MEDIAN_BAL_WITHOUT_DENY"))
+			Expect(parseErr).NotTo(HaveOccurred())
+
+			expectedError := constants.NewMismatchedOuterMethodsError(medianBalABI, medianBalWithoutDenyABI, "deny")
+			Expect(constants.CompareContractABI(medianBalABI, medianBalWithoutDenyABI)).To(MatchError(expectedError))
+		})
```
Comment (rmulhol) [inline pkg/shared/constants/external_test.go]: Looks like this test is the same as the above?

File: pkg/shared/constants/external.go [hunk 2]
```diff
@@ -140,27 +140,66 @@ func GetContractAddress(contract string) string {
...
-		return false
 	}
-OuterEvents:
 	for ak, av := range a.Events {
-		for bk, bv := range b.Events {
-			if ak == bk && av.String() == bv.String() {
-				continue OuterEvents
-			}
+		if av.String() != b.Events[ak].String() {
+			return NewMismatchedOuterEventsError(a, b, ak)
```
Comment (rmulhol) [inline pkg/shared/constants/external.go]: Maybe don't need `Outer` in the error name if there's no inner loop anymore?

File: pkg/shared/constants/external.go [hunk 3]
```diff
@@ -140,27 +140,66 @@ func GetContractAddress(contract string) string {
 	return getEnvironmentString("contract." + contract + ".address")
 }
 
-func parsedABIsAreEqual(a, b abi.ABI) bool {
+type MismatchedConstructorsError struct {
```
Comment (rmulhol) [inline pkg/shared/constants/external.go]: Wondering if we might want to drop these errors in a new/separate file

**Review by rmulhol — APPROVED** [review]
LGTM - excited to find errors more efficiently and describe them more clearly 🎉

File: pkg/shared/constants/external.go (hunk 2 above)
Comment (paytonrules) [reply pkg/shared/constants/external.go]: Heh - I thought Outer was a type of event.

File: pkg/shared/constants/external.go (hunk 3 above)
Comment (paytonrules) [reply pkg/shared/constants/external.go]: I like keeping the errors with the function, cause they are only used there, but I pulled the whole thing into its own file to address the noise.

I really just want a proper enum.

File: pkg/shared/constants/external_test.go (hunk 1 above)
Comment (paytonrules) [reply pkg/shared/constants/external_test.go]: 👍

**Pull request merged**